front: ensure destination is last in formatSuggestedViasToRowVias()

emersion opened the pull request:
See individual commits (some cleanups are also included).

See https://github.com/OpenRailAssociation/osrd/pull/8032#discussion_r1675411754

Note, the first commit has been moved from #8032 to this PR to avoid merge conflicts.

emersion committed:
front: simplify conditional in formatSuggestedViasToRowVias()

(-1 || 0) is always evaluated to -1.

We want to move the OP to the start of the array if we find it.
If it's already at the start, the logic inside the conditional is
a no-op.

emersion committed:
front: simplify swap in formatSuggestedViasToRowVias()

Instead of toSpliced() + unshift(), we can use array destructuring
to swap the first element with the one matching the origin.

emersion committed:
front: remove unnecessary cast in formatSuggestedViasToRowVias()

The array has the type PathStep[], so not need to cast.

emersion committed:
front: ensure destination is last in formatSuggestedViasToRowVias()

Apply the same logic as the origin, but for the destination.

See https://github.com/OpenRailAssociation/osrd/pull/8032#discussion_r1675411754

emersion commented:
I tried isolating the logic in a function to avoid repetition, but ended up with something that was IMHO harder to read. (Maybe could have a function to do the `findIndex()` call?)

SharglutDev commented on front/src/modules/timesStops/utils.ts:
I think your logic works great if origin/destination comes from the rocket but as we check only `if ('uic' in origin && 'ch' in origin)`, it won't work for an origin/destination added by map click (to test it, you can do a Paris St Lazare - Dieppe by clicking on map and see the destination at the wrong place).

emersion replied:
Do you think it would make sense to leverage `matchPathStepAndOp()` here? I was worried about breaking stuff since the checks are a bit different (`name` is checked, not `kp`).

SharglutDev replied:
I thought about your new util too which should be the best way to change this behavior. I just tested by putting `kp` instead of `name` and it seems to be working fine as well.
An easy test with multiple ops at position 0, is using a path "PNO AS". If you comment the origin switch logic, you will see the origin at the wrong place in the table and both fix with kp or name works.

emersion replied:
Done! See [c1f1b45](https://github.com/OpenRailAssociation/osrd/pull/8037/commits/c1f1b452d244bb6a78a203825d92afc916937547).

emersion committed:
front: generalize origin matching in formatSuggestedViasToRowVias()

We were only handling the case where the path step is a UIC. Use
matchPathStepAndOp() so that we handle all other cases (track
offset, trigram, OP ID) correctly too.

SharglutDev reviewed (approved):
Lgtm and tested
